plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitSource.java

```diff
@@ -362,6 +364,7 @@ else if (maxSplitBytes * 2 >= remainingBlockBytes) {
                     splitBytes = internalSplit.getEnd() - internalSplit.getStart();
                 }
 
+                OptionalLong initialRowId = OptionalLong.of((Long.MAX_VALUE / getBufferedInternalSplitCount()) * currentSplitIndex);
```

Comment:
Using `getBufferedInternalSplitCount()` for slicing id space between splits is not correct. It is not how many splits will there be in total for table. But how many ware currently buffered for execution.
Splits are dynamically loaded in the background by `BackgroundHiveSplitLoader` and value returned by `getBufferedInternalSplitCount()` will change over time. It also decreases as splits are taken away from the queue for execution.

I do not think we can do any better than what @electrum suggested in https://github.com/trinodb/trino/issues/8268#issuecomment-866180453:

> Allocate unique row IDs across writers by assigning ranges in the splits. For example, we could give each writer a large range of 2^42, which would allow both a huge number of rows and splits.

We need to statically decide how many splits we allow, and how many rows we allow for each split.
E.g. If we give each split 2^42 possible values we can have:
* 4,194,304 splits
* 4,398,046,511,104 rows per split

should be enough. Still we should control at runtime if we did not exceed either number of splits or number of rows generated for given split, and fail query nicely if that happens.

Reply:
Ok thanks for the in depth analysis, I will rework this slightly.